llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -5623,6 +5623,15 @@ bool SelectionDAG::isKnownNeverZero(SDValue Op, unsigned Depth) const {
   case ISD::ZERO_EXTEND:
   case ISD::SIGN_EXTEND:
     return isKnownNeverZero(Op.getOperand(0), Depth + 1);
+  case ISD::VSCALE: {
+    const Function &F = getMachineFunction().getFunction();
+    const APInt &Multiplier = Op.getConstantOperandAPInt(0);
+    ConstantRange CR =
+        getVScaleRange(&F, Op.getScalarValueSizeInBits()).multiply(Multiplier);
+    if (!CR.getUnsignedMin().isZero())
```

Comment:
I believe this check is incorrect.  Consider a wrapping range of the form [4, 2].  This includes 0 despite the 0 not being the minimum element.  The correct check here would be a contains check.

Reply:
Thanks for catching this, I was not aware ConstantRange could wrap. I've gone ahead and pushed a fix for this in baf22a527ca6571cdf5cd57c524a1fc261207947